proposals/NNNN-improve-VDE-for-stable-frame-rate.md

```diff
@@ -0,0 +1,165 @@
+                            // we can draw the image in Surface Texture here, something like
+                            long now = System.nanoTime();
+                            if (now > nextTime) {
+                                try {
+                                    windowSurface.makeCurrent();
+                                    GLES20.glViewport(0, 0, width, height);
+                                    fullFrameRect.drawFrame(textureId, matrix);
+                                } catch(RuntimeException e) {
+                                    ...
+                                }
+                                nextTime += frameIntervalInNano;
+                            }
+                            long delayTime = // adjust delaytime..
+                            handler.sendMessageDelayed(handler.obtainMessage(MSG_TICK), delayTime);
+                            break;
+                        case MSG_UPDATE_SURFACE:
+                            // update the surface here.
+                            updateSurface();
+                    }
+                }
+         }
+    }
+
+    private void updateSurface() {
+        try {
+            windowSurface.makeCurrent()
+        } catch(RuntimeException e) {
+            ...
+        }
+        surfacetexture.updateTexImage();
+        surfaceTexture.getTransformMatrix(matrix);
+    }
+
+    /**
+    * this is where we update the surface
+    * @param surfaceTexture
+    */
+    public void onFrameAvailable(SurfaceTexture surfaceTexture) {
+        // here, we can do update surfaceTexture
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            updateSurface();
+        } else {
+            // note that API level is lower than 21 (LOLLIPOP), setOnFrameAvailableListener(listener) is used,
+            // and most likely, OnFrameAvailableListener gets called in main thread.
+            // In that case, we have to call updateSurface in CaptureThread instead.
+            handler.sendMessage(handler.obtainMessafge(MSG_UPDATE_SURFACE));
+        }
+
+        // when the first time this gets called, start the loop
+        // by mHandler.sendMessage()
+        if (nextTime == 0) {
+            nextTime = System.nanoTime();
+            handler.sendMessage(handler.obtainMessage(MSG_TICK));
+        }
+    }
+ }
+```
+
+## Potential downsides
+
+No downside, as it will be implemented in Proxy, and there should be no performance overhead.
+One thing to note is that this proposal includes the usage of `Grafika` component, which is under [Apache License 2.0](http://www.apache.org/licenses/LICENSE-2.0).
+
+sdl_java_suite already uses some component under Apache License 2.0 (e.g. `JSON`), so adding another open source component won't cause any issues.
+
+## Impact on existing code
+
+Because this approach does not change existing API, and changes are inside of `VirtualDisplayEncoder` class, there's no impact to developers who use `VirtualDisplayEncoder`.
```

Comment:
```suggestion
Because this approach does not change an existing API, and changes are inside of `VirtualDisplayEncoder` class, there's no impact to developers who use `VirtualDisplayEncoder`.
```